### PR TITLE
python3Packages.gpib-ctypes: init at 0.3.0

### PR DESCRIPTION
**Pull request by fsagbuya**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kirillrdy** [comment]
Greetings ! and welcome.
please change commit messages

- maintainers: add fsagbuya
- pythonPackages.gpib-ctypes: init at 0.3.0

also update title of PR to pythonPackages.gpib-ctypes: init at 0.3.0

**Comment by fsagbuya** [comment]
> Greetings ! and welcome. please change commit messages
> 
> * maintainers: add fsagbuya
> * pythonPackages.gpib-ctypes: init at 0.3.0
> 
> also update title of PR to pythonPackages.gpib-ctypes: init at 0.3.0

Thank you. I already done these changes.

**Comment by fsagbuya** [comment]
> Also missing training new line at the end of the file,
> 
> https://github.com/NixOS/nixpkgs/actions/runs/4977647605/jobs/8906943716?pr=231989

I updated the commit. Thanks

**Comment by EvilMav** [comment]
Thank you for packaging this @fsagbuya, it is really useful in the lab =)

A few notes:
  * Not sure if mandatory, but it seems to be more customary to add `pytestCheckHook` to inputs rather then importing pytest and overriding `checkPhase` directly. 
  * I'm not sure `watchdog` is required as an input, it does not seem to be referenced in the sources...
  * You usually do not need to import `pip` directly but only `setuptools`

I'm new to python packaging in nix though, so take this with a grain of salt.

**Comment by EvilMav** [comment]
Actually, something else is probably missing. The package import `['libgpib.so.0', 'gpib-32.so']` in `gpib.py`. This has to come from somewhere. If you aren't running NixOS, it might be already available in your system and work, but there is a dependency missing here. You'd need to do two things:

  * identify and, if not packaged, package the native dependency and add it to buildInputs
  * patch `gpib.py` to include the path to the dependency withing the nix store in place of just `libgpib`. You'd probably use something like `substituteAll` for this.

**Comment by EvilMav** [comment]
> @EvilMav hi, can you explain how are you testing this ? nothing in nixpkgs is importing `gpib-ctypes`

I haven't yet tried testing, just eyeballing the code. It was actually on my TODO list for today to package for lab use ^^
It can be used optionally e.g. by `pyvisa-py` to access GPIB devices, so it should probably also go into `optional-dependencies` there.

**Comment by kirillrdy** [comment]
> > @EvilMav hi, can you explain how are you testing this ? nothing in nixpkgs is importing `gpib-ctypes`
> 
> I haven't yet tried testing, just eyeballing the code. It was actually on my TODO list for today to package for lab use ^^ It can be used optionally e.g. by `pyvisa-py` to access GPIB devices, so it should probably also go into `optional-dependencies` there.

greetings !

nothing in nixpkgs is importing gpid_ctypes so from nixpkgs point of view, its not being used, it needs to be added to `pkgs/top-level/python-packages.nix` so that it can be used in nixpkgs

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (kirillrdy) [inline pkgs/development/python-modules/gpib-ctypes/default.nix:30]: this needs to be a string, when trying to build this derivation

```
 error: URL literals are disabled

       at /home/kirillvr/.cache/nixpkgs-review/pr-231989-3/nixpkgs/pkgs/development/python-modules/gpib-ctypes/default.nix:30:16:

           29|     description = "Cross-platform Python bindings for the NI GPIB and linux-gpib C interfaces";
           30|     homepage = https://github.com/tivek/gpib_ctypes/;
             |                ^
           31|     license = licenses.gpl2Only;

```

Comment (fsagbuya) [reply pkgs/development/python-modules/gpib-ctypes/default.nix]: I missed this part. Thanks for noticing. Already updated it.

**Comment by kirillrdy** [comment]
@fsagbuya this derivation still doesnt build

maybe I suggest we mark this as draft until we resolve at least build issues

**Comment by fsagbuya** [comment]
> @fsagbuya this derivation still doesnt build
> 
> maybe I suggest we mark this as draft until we resolve at least build issues

Derivation already build.

**Comment by kirillrdy** [comment]
still doesn't build for me

```
nixpkgs-review pr 231989
$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/231989/head:refs/nixpkgs-review/1
$ git worktree add /home/kirillvr/.cache/nixpkgs-review/pr-231989-11/nixpkgs c3dc6acd1147fa4bb2da0147b00183d4a1ca2546
Preparing worktree (detached HEAD c3dc6acd114)
Updating files: 100% (35185/35185), done.
HEAD is now at c3dc6acd114 Merge pull request #232139 from r-ryantm/auto-update/toast
$ git merge --no-commit --no-ff 9d8f8c11765627f19a238e0915092877f95a4d71
Auto-merging maintainers/maintainer-list.nix
Auto-merging pkgs/top-level/python-packages.nix
Automatic merge went well; stopped before committing as requested
$ nix build --extra-experimental-features nix-command no-url-literals --no-link --keep-going --no-allow-import-from-derivation --option build-use-sandbox relaxed -f /home/kirillvr/.cache/nixpkgs-review/pr-231989-11/build.nix
error: hash mismatch in fixed-output derivation '/nix/store/fvxa5gv4vn8682xi1l7kl445bxp9sh6b-gpib_ctypes-0.3.0.tar.gz.drv':
         specified: sha256-D+UUxla1hvC4xOptIcK3NRdVPJeE2NFAcnYK/8Paka4=
            got:    sha256-c9l6TNmM4PtbvopnnFi5R1dQ9o3MI39BHHHPSGqfjCY=
error: 1 dependencies of derivation '/nix/store/xvcwxzia7p6w9rq9hspk2kh3f5v26m7f-python3.10-gpib_ctypes-0.3.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/9q9yyq34wak026h88bjd2r0n17vlk23a-python3.11-gpib_ctypes-0.3.0.drv' failed to build
error: 2 dependencies of derivation '/nix/store/0bi3wmmpdvkqj79cywgqrpiifd20lmqq-review-shell.drv' failed to build
```

**Comment by kirillrdy** [comment]
@ofborg build python3Packages.gpib-ctypes

**Comment by fsagbuya** [comment]
> still doesn't build for me
> 
> ```
> nixpkgs-review pr 231989
> $ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/231989/head:refs/nixpkgs-review/1
> $ git worktree add /home/kirillvr/.cache/nixpkgs-review/pr-231989-11/nixpkgs c3dc6acd1147fa4bb2da0147b00183d4a1ca2546
> Preparing worktree (detached HEAD c3dc6acd114)
> Updating files: 100% (35185/35185), done.
> HEAD is now at c3dc6acd114 Merge pull request #232139 from r-ryantm/auto-update/toast
> $ git merge --no-commit --no-ff 9d8f8c11765627f19a238e0915092877f95a4d71
> Auto-merging maintainers/maintainer-list.nix
> Auto-merging pkgs/top-level/python-packages.nix
> Automatic merge went well; stopped before committing as requested
> $ nix build --extra-experimental-features nix-command no-url-literals --no-link --keep-going --no-allow-import-from-derivation --option build-use-sandbox relaxed -f /home/kirillvr/.cache/nixpkgs-review/pr-231989-11/build.nix
> error: hash mismatch in fixed-output derivation '/nix/store/fvxa5gv4vn8682xi1l7kl445bxp9sh6b-gpib_ctypes-0.3.0.tar.gz.drv':
>          specified: sha256-D+UUxla1hvC4xOptIcK3NRdVPJeE2NFAcnYK/8Paka4=
>             got:    sha256-c9l6TNmM4PtbvopnnFi5R1dQ9o3MI39BHHHPSGqfjCY=
> error: 1 dependencies of derivation '/nix/store/xvcwxzia7p6w9rq9hspk2kh3f5v26m7f-python3.10-gpib_ctypes-0.3.0.drv' failed to build
> error: 1 dependencies of derivation '/nix/store/9q9yyq34wak026h88bjd2r0n17vlk23a-python3.11-gpib_ctypes-0.3.0.drv' failed to build
> error: 2 dependencies of derivation '/nix/store/0bi3wmmpdvkqj79cywgqrpiifd20lmqq-review-shell.drv' failed to build
> ```

I'll take a look on this. Will mark as draft for now.

**Comment by fsagbuya** [comment]
> xtra-experimental-feature

How can I test if it builds?

**Comment by kirillrdy** [comment]
> > xtra-experimental-feature
> 
> How can I test if it builds?

```
  git clone https://github.com/fsagbuya/nixpkgs
  cd nixpkgs/
  git checkout fsagbuya
  nix-build -A python3Packages.gpib-ctypes
   
```

```
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7ffff5877b80>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/pytest-runner/
ERROR: Could not find a version that satisfies the requirement pytest-runner (from versions: none)
ERROR: No matching distribution found for pytest-runner
Traceback (most recent call last):
  File "/nix/store/a9nrh0jkfxcf4v1arm0v8cg491wpi6y9-python3.10-setuptools-67.4.0/lib/python3.10/site-packages/setuptools/installer.py", line 97, in _fetch_build_egg_no_warn
    subprocess.check_call(cmd)
  File "/nix/store/95cxzy2hpizr23343b8bskl4yacf4b3l-python3-3.10.11/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/nix/store/95cxzy2hpizr23343b8bskl4yacf4b3l-python3-3.10.11/bin/python3.10', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/build/tmpk4n1m8z2', '--quiet', 'pytest-runner']' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/build/gpib_ctypes-0.3.0/nix_run_setup", line 8, in <module>
    exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\\r\\n', '\\n'), __file__, 'exec'))
  File "setup.py", line 28, in <module>
    setup(
  File "/nix/store/a9nrh0jkfxcf4v1arm0v8cg491wpi6y9-python3.10-setuptools-67.4.0/lib/python3.10/site-packages/setuptools/__init__.py", line 107, in setup
    _install_setup_requires(attrs)
  File "/nix/store/a9nrh0jkfxcf4v1arm0v8cg491wpi6y9-python3.10-setuptools-67.4.0/lib/python3.10/site-packages/setuptools/__init__.py", line 80, in _install_setup_requires
    _fetch_build_eggs(dist)
  File "/nix/store/a9nrh0jkfxcf4v1arm0v8cg491wpi6y9-python3.10-setuptools-67.4.0/lib/python3.10/site-packages/setuptools/__init__.py", line 85, in _fetch_build_eggs
    dist.fetch_build_eggs(dist.setup_requires)
  File "/nix/store/a9nrh0jkfxcf4v1arm0v8cg491wpi6y9-python3.10-setuptools-67.4.0/lib/python3.10/site-packages/setuptools/dist.py", line 894, in fetch_build_eggs
    return _fetch_build_eggs(self, requires)
  File "/nix/store/a9nrh0jkfxcf4v1arm0v8cg491wpi6y9-python3.10-setuptools-67.4.0/lib/python3.10/site-packages/setuptools/installer.py", line 39, in _fetch_build_eggs
    resolved_dists = pkg_resources.working_set.resolve(
  File "/nix/store/a9nrh0jkfxcf4v1arm0v8cg491wpi6y9-python3.10-setuptools-67.4.0/lib/python3.10/site-packages/pkg_resources/__init__.py", line 815, in resolve
    dist = self._resolve_dist(
  File "/nix/store/a9nrh0jkfxcf4v1arm0v8cg491wpi6y9-python3.10-setuptools-67.4.0/lib/python3.10/site-packages/pkg_resources/__init__.py", line 851, in _resolve_dist
    dist = best[req.key] = env.best_match(
  File "/nix/store/a9nrh0jkfxcf4v1arm0v8cg491wpi6y9-python3.10-setuptools-67.4.0/lib/python3.10/site-packages/pkg_resources/__init__.py", line 1123, in best_match
    return self.obtain(req, installer)
  File "/nix/store/a9nrh0jkfxcf4v1arm0v8cg491wpi6y9-python3.10-setuptools-67.4.0/lib/python3.10/site-packages/pkg_resources/__init__.py", line 1135, in obtain
    return installer(requirement)
  File "/nix/store/a9nrh0jkfxcf4v1arm0v8cg491wpi6y9-python3.10-setuptools-67.4.0/lib/python3.10/site-packages/setuptools/installer.py", line 99, in _fetch_build_egg_no_warn
    raise DistutilsError(str(e)) from e
distutils.errors.DistutilsError: Command '['/nix/store/95cxzy2hpizr23343b8bskl4yacf4b3l-python3-3.10.11/bin/python3.10', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/build/tmpk4n1m8z2', '--quiet', 'pytest-runner']' returned non-zero exit status 1.
/nix/store/37p8gq9zijbw6pj3lpi1ckqiv18j2g62-stdenv-linux/setup: line 1594: pop_var_context: head of shell_variables not a function context
error: builder for '/nix/store/djhp7xxv3m5sqxwpykrbi50a4p3dgjgv-python3.10-gpib_ctypes-0.3.0.drv' failed with exit code 1;
```

**Comment by fsagbuya** [comment]
> Actually, something else is probably missing. The package loads `['libgpib.so.0', 'gpib-32.so']` in `gpib/gpib.py`. This has to come from somewhere. If you aren't running NixOS, it might be already available in your system and work, but there is a dependency missing here. You'd need to do two things:
> 
> * identify and, if not packaged, package the native dependency and add it to buildInputs
> * patch `gpib.py` to include the path to the dependency withing the nix store in place of just `libgpib`. You'd probably use something like `substituteAll` for this.

Can you elaborate more on how to achieve this? (Also a newbie here) Current derivation is not building, maybe this is the cause. You can take a look at the error message on this thread.

**Comment by kirillrdy** [comment]
> > Actually, something else is probably missing. The package loads `['libgpib.so.0', 'gpib-32.so']` in `gpib/gpib.py`. This has to come from somewhere. If you aren't running NixOS, it might be already available in your system and work, but there is a dependency missing here. You'd need to do two things:
> > 
> > * identify and, if not packaged, package the native dependency and add it to buildInputs
> > * patch `gpib.py` to include the path to the dependency withing the nix store in place of just `libgpib`. You'd probably use something like `substituteAll` for this.
> 
> Can you elaborate more on how to achieve this? (Also a newbie here) Current derivation is not building, maybe this is the cause. You can take a look at the error message on this thread.

currently derivation is not building because of 

```
ERROR: Could not find a version that satisfies the requirement pytest-runner (from versions: none)
ERROR: No matching distribution found for pytest-runner
```

it can be fixed by adding `pytest-runner` to `checkInputs`

after that we will need to deal with missing libgpib

```
  /build/gpib_ctypes-0.3.0/gpib_ctypes/gpib/gpib.py:54: UserWarning: GPIB library not found. Please manually load it using _load_lib(filename). All GPIB functions will raise OSError until the library is manually loaded.

```

**Comment by kirillrdy** [comment]
i think we will need to package this first https://linux-gpib.sourceforge.io/

its currently not in nix

**Comment by fsagbuya** [comment]
> > > Actually, something else is probably missing. The package loads `['libgpib.so.0', 'gpib-32.so']` in `gpib/gpib.py`. This has to come from somewhere. If you aren't running NixOS, it might be already available in your system and work, but there is a dependency missing here. You'd need to do two things:
> > > 
> > > * identify and, if not packaged, package the native dependency and add it to buildInputs
> > > * patch `gpib.py` to include the path to the dependency withing the nix store in place of just `libgpib`. You'd probably use something like `substituteAll` for this.
> > 
> > 
> > Can you elaborate more on how to achieve this? (Also a newbie here) Current derivation is not building, maybe this is the cause. You can take a look at the error message on this thread.
> 
> currently derivation is not building because of
> 
> ```
> ERROR: Could not find a version that satisfies the requirement pytest-runner (from versions: none)
> ERROR: No matching distribution found for pytest-runner
> ```
> 
> it can be fixed by adding `pytest-runner` to `checkInputs`
> 
> after that we will need to deal with missing libgpib
> 
> ```
>   /build/gpib_ctypes-0.3.0/gpib_ctypes/gpib/gpib.py:54: UserWarning: GPIB library not found. Please manually load it using _load_lib(filename). All GPIB functions will raise OSError until the library is manually loaded.
> ```

Added `pytest-runner` and it builds with warning about the missing GPIB library. Need to add those next. Thanks!

**Comment by EvilMav** [comment]
@fsagbuya a few hints which might be helpful:

  * On issues of how to build and test stuff, I've found [this manual](https://nixos.wiki/wiki/Nixpkgs/Create_and_debug_packages) very helpful. To quickly build while testing, it is usually easiest to run `nix-build <YOUR_NIXPKGS_DIR_PATH> -A <PACKAGE_PATH>` 
  * it took me quite a while to figure out how to make the python packages actually find the binary ones. In normal executables Nix patches the `rpath` to make them load libraries from the store. The `ctypes` load libraries from default paths instead. It seems to be a common practice to patch the python files to replace the filenames of the `.so` libs to their full path in store. In the most cases in `nixpkgs`, this is done when applying a `.patch` file, which is preprocessed to contain the pathes in place of a `@some_lib@` placeholders. You can find many examples in `nixpkgs`, just grep for the word `patches` in `python-modules`. It can be easier though to just find and replace the paths if it is all you have to patch, unless the `nixpkgs` people object to this pattern. I have done this in some other (internal) package by adding this to `buildPythonPackage`:
```
postFixup = ''
        substituteInPlace $out/${python.sitePackages}/andor_emccd/*.py \
          --replace "\"/usr/local/etc/andor\"" "\"${andor-sdk}/etc/andor\"" \
          --replace "\"libandor.so\"" "\"${andor-sdk}/lib/libandor.so\""
      '';
```
where `substituteInPlace` is a function described in the manual, `python` which I take `site-packages` from has to be added to file's arguments, and `andor-sdk` here is the input you want to link to (in your case the linux-gpib).
Hope this saves you a bit of grey hair ;)

**Comment by fsagbuya** [comment]
Hello. Tried to build the package again and It does not return any error like before. It instead returns an entry of path to nix store of the package. 
```
/nix/store/n8fr07hdnqhzjl66v9axaja82g2pdr4g-python3.10-gpib_ctypes-0.3.0
```
Can you also verify on your machine? Thanks

**Comment by fsagbuya** [comment]
> Hello. Tried to build the package again and It does not return any error like before. It instead returns an entry of path to nix store of the package.
> 
> ```
> /nix/store/n8fr07hdnqhzjl66v9axaja82g2pdr4g-python3.10-gpib_ctypes-0.3.0
> ```
> 
> Can you also verify on your machine? Thanks

Please ignore this. The build error still exist when I update to the latest `nixpkgs`.

**Comment by fsagbuya** [comment]
@kirillrdy Hi. May I follow up on this PR? GPIB library can be manually loaded  depending on the user's specific requirements, don't you think?

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (kirillrdy) [inline pkgs/development/python-modules/gpib-ctypes/default.nix:21]: I was able to build this derivation without any of those nativeBuildInputs, are they necessary ?

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (kirillrdy) [inline pkgs/development/python-modules/gpib-ctypes/default.nix:32]: I suggest to add 

```
  pythonImportsCheck = [ "gpib_ctypes.gpib" ];
```

**Review by kirillrdy — COMMENTED** [review]
I personally don't object to merge this without gpib.

my understanding that this python library is not very useful without gpib.

If we merge this without porting gpib, we should also provide some indication to user how to manually load gpib

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (fsagbuya) [reply pkgs/development/python-modules/gpib-ctypes/default.nix:21]: Have able to build also without it. Will update the derivation.

**Comment by fsagbuya** [comment]
> If we merge this without porting gpib, we should also provide some indication to user how to manually load gpib
Do have suggestions on where to provide this?

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 231989` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.gpib-ctypes</li>
    <li>python310Packages.gpib-ctypes.dist</li>
    <li>python311Packages.gpib-ctypes</li>
    <li>python311Packages.gpib-ctypes.dist</li>
  </ul>
</details>

**Comment by Janik-Haag** [comment]
> > If we merge this without porting gpib, we should also provide some indication to user how to manually load gpib
> 
> Do have suggestions on where to provide this?

you can set the `meta.longDescription = ''''` attribute and describe the steps on how one would do that, this can be removed/shortand once gpib was packaged. 
Otherwise looks good to me and I'm fine with merging without gpib packaged.

**Comment by fsagbuya** [comment]
> you can set the `meta.longDescription = ''''` attribute and describe the steps on how one would do that, this can be removed/shortand once gpib was packaged. Otherwise looks good to me and I'm fine with merging without gpib packaged.

Thanks for the tip. Already updated it.

**Comment by Janik-Haag** [comment]
Okay actually I checked back with some other people, if you want to get this merged you have to package the other stuff as well so there are no imperative steps to the install.

**Comment by fsagbuya** [comment]
> Okay actually I checked back with some other people, if you want to get this merged you have to package the other stuff as well so there are no imperative steps to the install.

If I understand it right, there are no imperative steps for install here since loading the libraries is a user-intended method because it should match what device the user is testing it with. Those libraries could be proprietary as well, so using this package should already assume that the user has those libraries.

**Comment by Janik-Haag** [comment]
Take your time and feel free to ping me for review once your done :D

**Comment by fsagbuya** [comment]
> Take your time and feel free to ping me for review once your done :D

Hi @Janik-Haag. Invited you for review/merge of the linux-gpib package #245677. Other reviewer seems not responding.

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/gpib-ctypes/default.nix:32]:
```suggestion
```

This isn't need since nix doesn't run on windows, right? Or do you mean wsl?

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/gpib-ctypes/default.nix:35]: can you not package that for linux aswell and use patchelf on the library to fix the paths?

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/gpib-ctypes/default.nix:16]: you should add `disable = pythonOlder "3.XX";`

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (fsagbuya) [reply pkgs/development/python-modules/gpib-ctypes/default.nix:35]: Yes, package is pending at PR #245677 . Requesting for your help to merge.

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (fsagbuya) [reply pkgs/development/python-modules/gpib-ctypes/default.nix:32]: This description will be omitted once `linux-gpib` package is merge at PR #245677 and path is fixed.

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (fsagbuya) [reply pkgs/development/python-modules/gpib-ctypes/default.nix]: @Janik-Haag I have sent you an email regarding the `linux-gpib` package, requesting for your help to review and merge

Comment (Janik-Haag) [reply pkgs/development/python-modules/gpib-ctypes/default.nix]: I reviewed and merged the other PR. so we can omit this now. I'm also going to answer your email.

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/gpib-ctypes/default.nix:12]:
```suggestion
    version = "0.3.0";
    format = "...."
  
  disabled = pythonOlder "3.7";
```
please add the spacing and set the format, for more information about the format look at the manual in the `buildPythonPackage parameter` section https://nixos.org/manual/nixpkgs/stable/

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 231989` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.gpib-ctypes</li>
    <li>python310Packages.gpib-ctypes.dist</li>
    <li>python311Packages.gpib-ctypes</li>
    <li>python311Packages.gpib-ctypes.dist</li>
  </ul>
</details>

**Review by Janik-Haag — COMMENTED** [review]
```
/nix/store/gs91spjm51rah57d6hjb0fwzfhyi6p3s-python3.11-gpib_ctypes-0.3.0/lib/python3.11/site-packages/gpib_ctypes/gpib/gpib.py:54: UserWarning: GPIB library not found. Please manually load it using _load_lib(filename). All GPIB functions will raise OSError until the library is manually loaded.
```
when trying to import it throws this error, I guess you need a extra build input?

**Comment by fsagbuya** [comment]
> ```
> /nix/store/gs91spjm51rah57d6hjb0fwzfhyi6p3s-python3.11-gpib_ctypes-0.3.0/lib/python3.11/site-packages/gpib_ctypes/gpib/gpib.py:54: UserWarning: GPIB library not found. Please manually load it using _load_lib(filename). All GPIB functions will raise OSError until the library is manually loaded.
> ```
> 
> when trying to import it throws this error, I guess you need a extra build input?

Got the same warning before. The postFixup part should be able to resolve this though. It seems the library is required to be manually loaded on the user side.

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/gpib-ctypes/default.nix:11]:
```suggestion
  pname = "gpib-ctypes";
```
Please use a hyphen instead of an underscore.

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/gpib-ctypes/default.nix:24]:
```suggestion
```
`pytest-runner` is not needed, in fact.
We can remove this from setup.py in postPatch.

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/gpib-ctypes/default.nix:30]: There are other ways, such as using `os.add_dll_directory`, but I think this is OK.

```suggestion
    substituteInPlace gpib_ctypes/gpib/gpib.py \
      --replace "libgpib.so.0" "${linux-gpib}/lib/libgpib.so.0"
```
I think it could be simpler.

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (fsagbuya) [reply pkgs/development/python-modules/gpib-ctypes/default.nix:25]: How can I remove that from postPatch?

Comment (natsukium) [reply pkgs/development/python-modules/gpib-ctypes/default.nix:25]: with
```
substituteInPlace setup.py \
  --replace "'pytest-runner'," ""
```

Comment (fsagbuya) [reply pkgs/development/python-modules/gpib-ctypes/default.nix]: Updated the PR. Thanks

**Comment by natsukium** [comment]
LGTM.

Finally, would you change `pythonPackages` in the commit message and PR title to `python3Packages`?
Sorry for the late suggestion.

**Comment by natsukium** [comment]
@kirillrdy @evilmav @Janik-Haag 
Is there any suggestion?
I feel ready to merge.

**Comment by fsagbuya** [comment]
> LGTM.
> 
> Finally, would you change `pythonPackages` in the commit message and PR title to `python3Packages`? Sorry for the late suggestion.

Updated it. Thank you for the reviews.

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (kirillrdy) [inline pkgs/development/python-modules/gpib-ctypes/default.nix:39]: `changelog = "https://github.com/tivek/gpib_ctypes/blob/${version}/HISTORY.rst"`

**Review by natsukium — APPROVED** [review]

File: pkgs/development/python-modules/gpib-ctypes/default.nix
Comment (kirillrdy) [inline pkgs/development/python-modules/gpib-ctypes/default.nix:40]: haha sorry, need `;` at the end

**Review by kirillrdy — APPROVED** [review]

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 231989` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.gpib-ctypes</li>
    <li>python310Packages.gpib-ctypes.dist</li>
    <li>python311Packages.gpib-ctypes</li>
    <li>python311Packages.gpib-ctypes.dist</li>
  </ul>
</details>

**Review by Janik-Haag — APPROVED** [review]

**Comment by natsukium** [comment]
@ofborg python310Packages.gpib-ctypes

**Comment by natsukium** [comment]
ofborg doesn't seem to work well, but I can build it fine on aarch64-linux.

**Comment by kirillrdy** [comment]
@natsukium  ofborg stopped working for me too, but GrahamcOfBorg works

**Comment by fsagbuya** [comment]
Thank you all 😃

**Pull request merged**
